crates/dslab-mp/src/mc/strategy.rs

```diff
@@ -0,0 +1,213 @@
+
+        self.search_step_impl(system)?;
+
+        system.set_state(state);
+
+        Ok(())
+    }
+
+    fn apply_event(
+        &mut self,
+        system: &mut McSystem,
+        event_num: usize,
+        duplicate: bool,
+        corrupt: bool,
+    ) -> Result<(), String> {
+        let state = system.get_state(self.search_depth());
+
+        let mut event;
+        if duplicate {
+            event = self.duplicate_event(system, event_num);
+        } else {
+            event = self.take_event(system, event_num);
+        }
+
+        if corrupt {
+            event = self.corrupt_msg_data(event);
+        }
+
+        self.debug_log(&event, self.search_depth(), LogContext::Default);
+
+        system.apply_event(event);
+
+        self.search_step_impl(system)?;
+
+        system.set_state(state);
+
+        Ok(())
+    }
+
+    fn take_event(&self, system: &mut McSystem, event_num: usize) -> McEvent {
+        system.events.remove(event_num)
+    }
+
+    fn clone_event(&self, system: &mut McSystem, event_num: usize) -> McEvent {
+        system.events[event_num].clone()
+    }
+
+    fn add_event(&self, system: &mut McSystem, event: McEvent, event_num: usize) {
+        system.events.insert(event_num, event);
+    }
+
+    fn corrupt_msg_data(&self, event: McEvent) -> McEvent {
+        self.debug_log(&event, self.search_depth(), LogContext::Corrupted);
+        match event {
+            MessageReceived {
+                msg,
+                src,
+                dest,
+                options,
+            } => {
+                lazy_static! {
+                    static ref RE: Regex = Regex::new(r#""\w+""#).unwrap();
+                }
+                let corrupted_data = RE.replace_all(&msg.data, "\"\"").to_string();
+                MessageReceived {
+                    msg: Message::new(msg.tip, corrupted_data),
+                    src,
+                    dest,
+                    options,
+                }
+            }
+            _ => event,
+        }
+    }
+
+    fn duplicate_event(&self, system: &mut McSystem, event_num: usize) -> McEvent {
+        let event = self.take_event(system, event_num);
+        self.add_event(system, event.duplicate().unwrap(), event_num);
+        self.debug_log(&event, self.search_depth(), LogContext::Duplicated);
+        event
+    }
+
+    fn debug_log(&self, event: &McEvent, depth: u64, log_context: LogContext) {
+        if self.log_mode() == &LogMode::Debug {
+            match event {
+                MessageReceived { msg, src, dest, .. } => {
+                    self.log_message(depth, msg, src, dest, log_context);
+                }
+                TimerFired { proc, timer } => {
+                    t!(format!("{:>10} | {:>10} !-- {:<10}", depth, proc, timer).yellow());
+                }
+                _ => {
+                    t!(format!("Internal error: unknown event in model checking Strategy").red());
+                }
```

Comment:
Паниковать надо раньше в process_event

```suggestion
                _ => {}
```